### PR TITLE
Updated parsing of ModelStructure

### PR DESCRIPTION
**Pull request by PeterMeisrimelModelon**

### CODE REVIEW

File: src/XML/src/FMI3/fmi3_xml_model_structure.c [hunk 1]
```diff
@@ -442,67 +406,69 @@ int fmi3_xml_parse_unknown(fmi3_xml_parser_context_t *context,
         return -1;
     }
 
-    return fmi3_xml_parse_dependencies(context, parentElmID, deps);
+    return fmi3_xml_parse_dependencies(context, elmID, deps);
 }
 
 
-/*int fmi3_xml_handle_OutputUnknown(fmi3_xml_parser_context_t *context, const char* data) {*/
+// TODO: Currently required for structural purposes, but not a valid ModelStructure element on its own
```
Comment (PeterMeisrimelModelon) [inline src/XML/src/FMI3/fmi3_xml_model_structure.c:413]: This is something that might be desirable to fix here, not sure what would be the best way.

File: Test/FMI3/fmi3_import_model_structure_test.c [hunk 2]
```diff
@@ -92,7 +92,43 @@ static int test_parse_xml(int expectFailure, char* xmlTestRootDir, char* xmlDirR
...
+static void logger_incorrect_order(jm_callbacks* cb, jm_string module,
+        jm_log_level_enu_t log_level, jm_string message)
+{
+    char* expMsg = "Parse error at line";
+    if (!strncmp(expMsg, message, strlen(expMsg))) {
+        g_logger_found_err_msg = 1;
+    }
+
+    jm_default_logger(cb, module, log_level, message);
+}
```
Comment (filip-stenstrom) [inline Test/FMI3/fmi3_import_model_structure_test.c:113]: If you rebase after #44 you can use fmi_testutil functions to find any log message.

File: Test/FMI3/fmi3_import_model_structure_test.c [hunk 3]
```diff
@@ -92,7 +92,43 @@ static int test_parse_xml(int expectFailure, char* xmlTestRootDir, char* xmlDirR
...
+
+/* Tests that an error is raised when for outputs with incorrect causality. */
+static void logger_output_wrong_causality(jm_callbacks* cb, jm_string module,
+        jm_log_level_enu_t log_level, jm_string message)
+{
+    char* expMsg = "TODO";
```
Comment (filip-stenstrom) [inline Test/FMI3/fmi3_import_model_structure_test.c:131]: Reminder

File: Test/FMI3/parser_test_xmls/model_structure/invalid/fmi2_style_list/modelDescription.xml [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  fmiVersion="3.0"
+  modelName="myModelName"
+  guid="myGuid"
+  description="myDescription"
+  author="myAuthor"
+  generationTool="myGenerationTool"
+  generationDateAndTime="2002-05-30T09:00:00"
+  version="myModelVersion"
+  variableNamingConvention="structured"
+  numberOfEventIndicators="3">
```
Comment (filip-stenstrom) [inline Test/FMI3/parser_test_xmls/model_structure/invalid/fmi2_style_list/modelDescription.xml:12]: I think we should remove all optional attributes to minimize clutter. You can copy from #44. I'm trying to do it for all new XMLs.

File: Test/FMI3/parser_test_xmls/model_structure/invalid/incorrect_order/modelDescription.xml [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+  modelName="myModelName"
+  instantiationToken="myInstToken">
+
+<ModelExchange
+  modelIdentifier="myModelIdentifier" />
+
+<ModelVariables>
+    <Float64 name="state_var" valueReference="0"/>
+    <Float64 name="der(state_var)" valueReference="1" causality="output" variability="continuous" initial="approx" derivative="1" start="1.0"/>
+    <Float64 name="output_var" valueReference="2"/>
```
Comment (filip-stenstrom) [inline Test/FMI3/parser_test_xmls/model_structure/invalid/incorrect_order/modelDescription.xml:13]: causality=output should be on output_var

File: Test/FMI3/parser_test_xmls/model_structure/valid/modelDescription.xml [hunk 6]
```diff
@@ -10,11 +10,13 @@
 <ModelVariables>
     <Float64 name="state_var" valueReference="0"/>
     <Float64 name="der(state_var)" valueReference="1" causality="output" variability="continuous" initial="approx" derivative="1" start="1.0"/>
+    <Float64 name="output_var" valueReference="2" causality="output"/>
+    <Float64 name="init_unknown_var" valueReference="3"/>
```
Comment (filip-stenstrom) [inline Test/FMI3/parser_test_xmls/model_structure/valid/modelDescription.xml:14]: This is a local variable. They're not considered InitialUnknown right?

File: src/CAPI/include/FMI3/fmi3_capi.h [hunk 7]
```diff
@@ -728,14 +728,14 @@ fmi3_status_t fmi3_capi_completed_integrator_step(fmi3_capi_t* fmu,
...
  *
  * @param fmu C-API struct that has succesfully loaded the FMI function.
- * @param derivatives (Output) Array of the derivatives.
- * @param nx Number of derivatives.
+ * @param derivatives (Output) Array of the continuous state derivatives.
+ * @param nx Number of continuous state derivatives.
  * @return FMI status.
  */
-fmi3_status_t fmi3_capi_get_derivatives(fmi3_capi_t* fmu, fmi3_float64_t derivatives[], size_t nx);
+fmi3_status_t fmi3_capi_get_continuous_state_derivatives(fmi3_capi_t* fmu, fmi3_float64_t continuousStateDerivatives[], size_t nx);
```
Comment (filip-stenstrom) [inline src/CAPI/include/FMI3/fmi3_capi.h:738]: Good you found this.

We probably want the args to match here and in the impl (.c file), so I suggest naming continuousStateDerivatives just `derivatives` again. It's quite clear what it is, and it's the same as in the standard header

File: src/XML/src/FMI3/fmi3_xml_model_structure.c (hunk 1 above)
Comment (filip-stenstrom) [reply src/XML/src/FMI3/fmi3_xml_model_structure.c:413]: Add an assert(0) as well and we're good.

For some other elements that can appear on multiple places, the one that happens first in the XML (e.g. Real under TypeDefinitions in FMI2), is the handler for that element (..._handle_Real), while the ones that happen later need to name something like (..._handle_RealVariable). I think that's more confusing.

I don't think it's worth the effort trying to rewrite it

File: src/XML/src/FMI3/fmi3_xml_model_structure.c [hunk 8]
```diff
@@ -442,67 +406,69 @@ int fmi3_xml_parse_unknown(fmi3_xml_parser_context_t *context,
...
+    if(!data) {
+
         fmi3_xml_model_description_t* md = context->modelDescription;
         fmi3_xml_model_structure_t* ms = md->modelStructure;
         fmi3_xml_variable_t* derXX; /* float64 or float32 variable */
         int validDeriv; /* valid derivative found */
 
         /* perform the parsing */
-        int status = fmi3_xml_parse_unknown(context, fmi3_xml_elmID_Derivatives, &ms->derivatives, ms->derivativeDeps);
+        int status = fmi3_xml_parse_unknown(context, fmi3_xml_elmID_ContinuousStateDerivative, &ms->continuousStateDerivatives, ms->continuousStateDerivativeDeps);
```
Comment (filip-stenstrom) [inline src/XML/src/FMI3/fmi3_xml_model_structure.c:439]: long line

**Review by filip-stenstrom — COMMENTED** [review]
Many TODOs but given this is first subtask that's fine

File: src/XML/src/FMI3/fmi3_xml_model_structure.c [hunk 9]
```diff
@@ -232,12 +219,14 @@ int fmi3_xml_parse_dependencies(fmi3_xml_parser_context_t *context,
...
             if(!ch) break;
             if(sscanf(cur, "%d", &ind) != 1) {
+                // TODO: Rework this error; list actual type
                 fmi3_xml_parse_error(context, "XML element 'Unknown': could not parse item %d in the list for attribute 'dependencies'",
                     numDepInd);
                ms->isValidFlag = 0;
                return 0;
             }
             if(ind < 1) {
+                // TODO: Rework this error; list actual type
```
Comment (PeterMeisrimelModelon) [inline src/XML/src/FMI3/fmi3_xml_model_structure.c:229]: Multiple occasions of this one, ties into the same issue as discussed in https://github.com/modelon-community/fmi-library/pull/44#discussion_r1185144221

File: src/XML/src/FMI3/fmi3_xml_model_structure.c (hunk 1 above)
Comment (PeterMeisrimelModelon) [reply src/XML/src/FMI3/fmi3_xml_model_structure.c]: Resolved this by instead removing the fmi3_xml_scheme_Unknown instead. It was essentially an abstract type and that behavior is somewhat kept by having an internal shared parsing function.

Note: Unknown elements in ModelStructure are now skipped (with log message).

File: Test/FMI3/fmi3_import_model_structure_test.c (hunk 3 above)
Comment (PeterMeisrimelModelon) [reply Test/FMI3/fmi3_import_model_structure_test.c]: Fixed.

File: Test/FMI3/parser_test_xmls/model_structure/invalid/fmi2_style_list/modelDescription.xml (hunk 4 above)
Comment (PeterMeisrimelModelon) [reply Test/FMI3/parser_test_xmls/model_structure/invalid/fmi2_style_list/modelDescription.xml]: Fixed. I had simply copy pasted the valid one and then broke it, didn't consider the overhead here.

File: Test/FMI3/parser_test_xmls/model_structure/valid/modelDescription.xml (hunk 6 above)
Comment (PeterMeisrimelModelon) [reply Test/FMI3/parser_test_xmls/model_structure/valid/modelDescription.xml]: Fixed, also with error check on this now.

File: src/CAPI/include/FMI3/fmi3_capi.h (hunk 7 above)
Comment (PeterMeisrimelModelon) [reply src/CAPI/include/FMI3/fmi3_capi.h]: Changed the arg to derivatives for consistency (and shorter lines)

File: src/XML/src/FMI3/fmi3_xml_model_structure.c [hunk 10]
```diff
@@ -55,12 +55,12 @@ void fmi3_xml_free_model_structure(fmi3_xml_model_structure_t* ms) {
     cb = ms->outputs.callbacks;
 
     jm_vector_free_data(jm_voidp)(&ms->outputs);
-    jm_vector_free_data(jm_voidp)(&ms->derivatives);
+    jm_vector_free_data(jm_voidp)(&ms->continuousStateDerivatives);
     jm_vector_free_data(jm_voidp)(&ms->discreteStates);
```
[the remark this answers is not in the document]
Comment (PeterMeisrimelModelon) [reply src/XML/src/FMI3/fmi3_xml_model_structure.c]: Fixed, please have a second look that I didn't miss anything here for FMI3.

File: Test/FMI3/parser_test_xmls/model_structure/invalid/incorrect_order/modelDescription.xml (hunk 5 above)
Comment (PeterMeisrimelModelon) [reply Test/FMI3/parser_test_xmls/model_structure/invalid/incorrect_order/modelDescription.xml]: Fixed, also with error check on this now.

File: Test/FMI3/parser_test_xmls/model_structure/valid/modelDescription.xml (hunk 6 above)
Comment (PeterMeisrimelModelon) [reply Test/FMI3/parser_test_xmls/model_structure/valid/modelDescription.xml]: Right. Would require initial="calculated" or similar. Removed it for simplicity here and added the derivative as InitialUnknown instead (which is also required).

File: src/XML/src/FMI3/fmi3_xml_model_structure.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (filip-stenstrom) [reply src/XML/src/FMI3/fmi3_xml_model_structure.c]: Great. Most important is to get rid of API (which is done). If we find anything later we can take it then

[the remark this answers is not in the document]
Comment (filip-stenstrom) [reply src/XML/src/FMI3/fmi3_xml_model_structure.c]: You can resolve this

File: Test/FMI3/fmi3_import_model_structure_test.cpp [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+
+    fmi3_testutil_import_free(tfmu);
+}
+
+TEST_CASE("Warning check: ModelStructure; FMI2 style lists") {
+    const char* xmldir = FMI3_TEST_XML_DIR "/model_structure/invalid/fmi2_style_list";
+
+    fmi3_testutil_import_t* tfmu = fmi3_testutil_parse_xml_with_log(xmldir);
+    REQUIRE(tfmu != nullptr);
+    REQUIRE(tfmu->fmu != nullptr); // Parsing is successfull
```
Comment (filip-stenstrom) [inline Test/FMI3/fmi3_import_model_structure_test.cpp:108]:
```suggestion
    REQUIRE(tfmu->fmu != nullptr); // Parsing is successful
```

File: Test/FMI3/fmi3_import_model_structure_test.cpp [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+
+    fmi3_import_variable_list_t* varList = fmi3_import_get_initial_unknowns_list(fmu);
+
+    size_t nVar = 1;
+    REQUIRE(fmi3_import_get_variable_list_size(varList) == nVar);
+
+    fmi3_import_variable_t* var = fmi3_import_get_variable(varList, 0);
+
+    fmi3_value_reference_t vr = 1;
+    REQUIRE(fmi3_import_get_variable_vr(var) == vr);
```
Comment (filip-stenstrom) [inline Test/FMI3/fmi3_import_model_structure_test.cpp:51]: These kind of variables I'd just inline to make the tests more compact, but up to you. (I know that on test failure you'll get the variable name unexpanded, but it's quite obvious that we're comparing with a VR)

**Review by filip-stenstrom — APPROVED** [review]
Good that you converted the test and changed to new testutil!

Minor optional fixes.

File: Test/FMI3/fmi3_import_model_structure_test.cpp (hunk 11 above)
Comment (PeterMeisrimelModelon) [reply Test/FMI3/fmi3_import_model_structure_test.cpp]: Fixed, also a couple other occasions.

File: Test/FMI3/fmi3_import_model_structure_test.cpp (hunk 12 above)
Comment (PeterMeisrimelModelon) [reply Test/FMI3/fmi3_import_model_structure_test.cpp]: Fixed, made the tests a bit more compact.

File: src/XML/src/FMI3/fmi3_xml_model_structure.c (hunk 9 above)
Comment (PeterMeisrimelModelon) [reply src/XML/src/FMI3/fmi3_xml_model_structure.c:230]: Not fixing this in this MR.

**Pull request merged**